---
title: "Annotation Report"
author: "`r params$query` genome assembly"
date: "`r format(Sys.time(), '%d %B %Y')`"
subtitle: "Virulence genes annotated using VFDB and Victors databases"
params:
  blast_id: na
  blast_cov: na
  vfdb_blast: na
  victors_blast: na
  query: na
  gff: na
output:
  bookdown::html_document2:
    df_print: paged
    number_sections: false
    syntax: tango
    self_contained: yes
    theme: united
---

<style type="text/css">

body{ /* Normal  */
      font-size: 14px;
  }
.main-container {
  max-width: 85% !important;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
suppressMessages(library(magrittr))
suppressMessages(library(knitr))
suppressMessages(library(kableExtra))
suppressMessages(library(stringr))
suppressMessages(library(ballgown))
suppressMessages(library(tidyr))
suppressMessages(library(plyr))
suppressMessages(library(DT))
suppressMessages(library(dplyr))
suppressMessages(library(stringr))
suppressMessages(library(ggplot2))

victors_blast <- try(read.delim(params$victors_blast, header = FALSE), silent = TRUE)
vfdb_blast <- try(read.delim(params$vfdb_blast, header = FALSE), silent = TRUE)
gff <- try(gffRead(params$gff), silent = TRUE)

# Check if blast is empty
if (class(vfdb_blast) != 'try_error' && !is.null(nrow(vfdb_blast)) > 0) {
  vfdb_not_null <- TRUE
  vfdb_null <- FALSE
} else {
  vfdb_not_null <- FALSE
  vfdb_null <- TRUE
}

if (class(victors_blast) != 'try_error' && !is.null(nrow(victors_blast)) > 0) {
  victors_not_null <- TRUE
  victors_null <- FALSE
} else {
  victors_not_null <- FALSE
  victors_null <- TRUE
}

# Function to get Attribute Fields
getAttributeField <- function (x, field, attrsep = ";") { 
  s = strsplit(x, split = attrsep, fixed = TRUE) 
  sapply(s, function(atts) { 
    a = strsplit(atts, split = "=", fixed = TRUE) 
    m = match(field, sapply(a, "[", 1)) 
    if (!is.na(m)) { rv = a[[m]][2] 
    } 
    else { 
      rv = as.character(NA) 
    } 
    return(rv) 
  }) 
}
```

# About

Virulence factors are molecules produced by bacteria that add effectiveness in their colonization of a niche, immunoevasion, immunosuppression and obtain nutrition from the host.

[VFDB](http://www.mgc.ac.cn/VFs/main.htm) (Virulence Factor Database) is a comprehensive resource, created in 2004, of curated information about virulence factors of pathogenic bacteria. To date, it contains a 1080 virulence factors in its database, from 74 _bacteria genera_.

[Victors](http://www.phidias.us/victors/) is a curated database which currently possesses 5296 virulence factors for 194 different pathogens including bacteria, viruses and parasites.

[VFDB](http://www.mgc.ac.cn/VFs/main.htm) and [Victors](http://www.phidias.us/victors/) databases were blasted via BLASTx, using the thresholds below. This report summarizes its results.

* Blast % identity: `> `r params$blast_id`%`
* Blast % query coverage: `> `r params$blast_cov`%`

# Results {.tabset .tabset-fade .tabset-pills}

## VFDB

```{r, VFDB_conditional_block, echo=FALSE, results='asis', eval=vfdb_not_null, child='yes_vfdb.Rmd'}
```

```{r, VFDB_conditional_block_2, echo=FALSE, results='asis', eval=vfdb_null, child='no_vfdb.Rmd'}
```

## Victors

```{r, Victors_conditional_block, echo=FALSE, results='asis', eval=victors_not_null, child='yes_victors.Rmd'}
```

```{r, Victors_conditional_block_2, echo=FALSE, results='asis', eval=victors_null, child='no_victors.Rmd'}
```