---
title: "Annotation Report of query genome `r params$query`"
author: "Felipe M. Almeida <falmeida@aluno.unb.br>"
date: "`r format(Sys.time(), '%d %B %Y')`"
subtitle: "Prophages summary."
params:
  blast_id:
  blast_cov:
  phigaro_dir:
  phigaro_txt:
  virsorter_dir:
  virsorter_csv:
  phast_blast:
  query:
output:
  bookdown::html_document2:
    df_print: paged
    number_sections: false
    syntax: tango
    self_contained: yes
    theme: journal
---

<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
options(knitr.table.format = "html")
bt_opt =  c("hover", "condensed", "responsive")
library(magrittr)
library(knitr)
library(tidyr)
library(plyr)
library(kableExtra)
library(dplyr)
library(stringr)
library(ggplot2)
phigaro_dir <- params$phigaro_dir
phigaro_html <- paste0(phigaro_dir, "/assembly.phg.html", collapse = "")
phigaro_txt <- params$phigaro_txt
virsorter_dir <- params$virsorter_dir
virsorter_csv <- params$virsorter_csv

## Read phigaro
phigaro_txt <- try(read.delim(phigaro_txt), silent = TRUE)
if (class(phigaro_txt) != 'try_error' & nrow(phigaro_txt) > 0) {
  phigaro_not_null <- TRUE
  phigaro_null <- FALSE
} else {
  phigaro_not_null <- FALSE
  phigaro_null <- TRUE
}

## Read virsorter
virsorter_csv <- try(read.csv(virsorter_csv, comment.char = "#", header = FALSE, col.names = c("Contig_id", "Nb genes contigs", "Fragment", "Nb genes", "Category", "Nb phage hallmark genes", "Phage gene enrichment sig", "Non-Caudovirales phage gene enrichment sig", "Pfam depletion sig", "Uncharacterized enrichment sig", "Strand switch depletion sig", "Short genes enrichment sig")), silent = TRUE)
if (class(virsorter_csv) != 'try_error' & nrow(virsorter_csv) > 0) {
  virsorter_not_null <- TRUE
  virsorter_null <- FALSE
} else {
  virsorter_not_null <- FALSE
  virsorter_null <- TRUE
}

## Read PHAST
phast_blast <- try(read.delim(params$phast_blast, header = FALSE, col.names = c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "slen", "evalue", "bitscore", "stitle")), silent = TRUE)
if (class(phast_blast) != 'try_error' & nrow(phast_blast) > 0) {
  phast_not_null <- TRUE
  phast_null <- FALSE
} else {
  phast_not_null <- FALSE
  phast_null <- TRUE
}
```

# About

Prophage is a bacteriophage that have been inserted and integrated into the bacterial chromosome or plasmid. It is the latent form of a phage.

Prophages were annotated using [PHAST](https://phaster.ca) database, [virsorter](https://github.com/simroux/VirSorter) and [phigaro](https://github.com/bobeobibo/phigaro) software. PHAST database was searched with BLAST to annotate all genes found in the query genome that are known to be found in prophages.

Blast thresholds were the following:

* Blast % identity: `> `r params$blast_id`%`
* Blast % query coverage: `> `r params$blast_cov`%`

Additionally, virsorter and phigaro software are used to identify the genomic regions of prophages within the query genome and to extract its DNA sequences.

Therefore, this report is a combination of identified prophage's genomic regions (using virsorter and phigaro) and its gene functional annotation (using PHAST database).

# Results

## Phigaro

```{r, phigaro_conditional_block_TRUE, echo=FALSE, results='asis', eval=phigaro_not_null, child='yes_phigaro.Rmd'}
```

```{r, phigaro_conditional_block_FALSE, echo=FALSE, results='asis', eval=phigaro_null, child='no_phigaro.Rmd'}
```

## Virsorter

```{r, phigaro_conditional_block_TRUE, echo=FALSE, results='asis', eval=virsorter_not_null, child='yes_virsorter.Rmd'}
```

```{r, phigaro_conditional_block_FALSE, echo=FALSE, results='asis', eval=virsorter_null, child='no_virsorter.Rmd'}
```

## PHAST

```{r, phigaro_conditional_block_TRUE, echo=FALSE, results='asis', eval=phast_not_null, child='yes_phast.Rmd'}
```

```{r, phigaro_conditional_block_FALSE, echo=FALSE, results='asis', eval=phast_null, child='no_phast.Rmd'}
```