---
title: "Antimicrobial Resistance Genes Report"
params:
  resistance_summary:
  is_CARD:
  card_summary:
  ncbi_amr:
output:
  bookdown::html_document2:
    df_print: paged
    number_sections: false
    syntax: tango
    self_contained: yes
    theme: journal
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
options(knitr.table.format = "html")
bt_opt =  c("hover", "condensed", "responsive")
library(magrittr)
library(knitr)
library(tidyr)
library(plyr)
library(kableExtra)
library(dplyr)
library(stringr)
library(ggplot2)
resistance_summary <- read.delim(params$resistance_summary)
ncbi_amr <- read.delim(params$ncbi_amr)
is_CARD <- params$is_CARD
card_summary <- read.delim(params$card_summary)
```

# About

This report contains all the resistance related genes found when searching the genome with [CARD](https://card.mcmaster.ca/), [NCBI-amr](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA313047) and [Resfams](http://www.dantaslab.org/resfams) datases. Additionally, Resfams hmm file have been addded to Prokka datases, thus, this report also contains a general report of non drug related resistance genes.

# NCBI-amr

NCBI has produced a hmm file of curated AMR genes. This database was incorporated into Prokka databases. Therefore, we are able to easily retrieve genes annotated from this curated dataset.

This dataset is not exclusive to drug related genes and it contains several types of resistance factors such as resistance against heavy metals. Therefore, it not possible to create an antibiogram from this dataset. However, all resistance genes from this dataset and its protein ID (local protein ID) are shown in Table \@ref(tab:ncbi-amr-resistance-genes).

# Resfams

These elements were annotated from the [Resfams database](http://www.dantaslab.org/resfams/). This database contains genes related to antibiotic resistance genes as well as genes related to environmental stress resistance. Similar to NCBI-amr results, we can't create an antibiogram using this database but it is extremely informative in terms of environmental resistance to stress. This data is summarised in Table \@ref(tab:resfams-resistance-table).

```{r, CARD_conditional_block, echo=FALSE, results='asis', eval=is_CARD, child='CARD.Rmd'}
```

```{r ncbi-amr-resistance-genes}
genes <- ncbi_amr %>% select(Prokka_ID, Prokka_product, Prokka_inference)
genes <- separate(genes, Prokka_inference, c("method", "prog", "ver", "motif", "db_id"), sep = ":", extra = "merge")
genes <- genes %>% select("Prokka_ID", "Prokka_product", "db_id")

# Create urls
urls <- lapply(genes$db_id, function(x){paste0("https://www.ncbi.nlm.nih.gov/protein/?term=", x)})

# Produce Table
genes %>%
  mutate("db_id" = cell_spec(db_id, "html", link = urls)) %>%
  arrange(Prokka_product) %>%
kable(caption = "Resistance genes annotated from NCBI-amr curated database", escape = FALSE, col.names = c("Local Protein ID", "Annotated Gene Product", "Reference Motif")) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 10) %>%
  collapse_rows(columns = 1:3, valign = "middle") %>%
  footnote(general = "The last column contains a link to NCBI showing the resulting protein. Note that, the result shows all species that have that protein annotated using this motif. It is not a similarity search.")
```

```{r resfams-resistance-table}
resistance_summary %>%
  filter(str_detect(Prokka_inference, "resfams")) %>%
  select(Prokka_ID, Prokka_product) %>%
  kable(caption = "Resistance genes annotated from Resfams database", escape = FALSE, col.names = c("Local Protein ID", "Annotated Gene Product")) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 10)
```

```{r card-resistance-pattern, eval=is_CARD}
# Create urls
urls <- lapply(card_summary$CVTerm, function(x){paste0("https://card.mcmaster.ca/ontology/", x)})

# Produce Table
card_summary %>%
  select("Name", "Drug_Class", "Resistance_Mechanism", "ARO_Accession", "CVTerm") %>%
  mutate("CVTerm" = cell_spec(CVTerm, "html", link = urls)) %>%
  separate_rows(Drug_Class, sep = ":") %>%
  group_by(Name) %>%
  kable(caption = "CARD resistance genes annotated and its target drug classes", escape = FALSE) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 10) %>%
  collapse_rows(columns = 1:5, valign = "middle") %>%
  footnote(general = "CVTerms are the gene IDs that direct link the results to CARD database")
```