---
title: "Annotation Report"
author: "Felipe M. Almeida <falmeida@aluno.unb.br>"
date: "`r format(Sys.time(), '%d %B %Y')`"
subtitle: "Prophages summary."
params:
  blast_id:
  blast_cov:
  phigaro_dir:
  phigaro_txt:
  phigaro_html:
  virsorter_dir:
  virsorter_csv:
output:
  bookdown::html_document2:
    df_print: paged
    number_sections: false
    syntax: tango
    self_contained: yes
    theme: journal
---

<style type="text/css">

body{ /* Normal  */
      font-size: 16px;
  }
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
options(knitr.table.format = "html")
bt_opt =  c("hover", "condensed", "responsive")
library(magrittr)
library(knitr)
library(tidyr)
library(plyr)
library(kableExtra)
library(dplyr)
library(stringr)
library(ggplot2)
phigaro_dir <- params$phigaro_dir
phigaro_html <- params$phigaro_html
phigaro_txt <- params$phigaro_txt
virsorter_dir <- params$virsorter_dir
virsorter_csv <- params$virsorter_csv
```

# About

Prophages were annotated using PHAST database, virsorter and phigaro software. PHAST database was serched with BLAST to annotate all genes that are know to be found in prophages and were also found in query genome.

Additionally, virsorter and phigaro software are used to identify the genomic regions of prophages within the query genome and to extract its DNA sequences.

Therefore, this report is a combination of identified genomic regions of prophages using virsorter and phigaro with the gene functional annotation using PHAST database.

# Prophages

## Phigaro

Genomic regions identyfied as prophages by the software phigaro are all showed in its [html report file](`r phigaro_html`) stored at ``r phigaro_dir``. Its genomic regions are summarised in Table \@ref(tab:phigaro-table)

```{r phigaro-table}
phigaro_txt <- read.delim(phigaro_txt)
# Produce Table
phigaro_txt %>%
  kable(caption = "Prophages annotated with phigaro software", escape = FALSE) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 12)
```

## Virsorter

Outputs of virsorter software are in ``r virsorter_dir`` directory. Within this directory, all predicted prophages are stored in a single fasta file at `fasta` dir. A summary of virsorter annotatation is showed in Table \@ref(tab:virsorter-table). All features can fall within three categories. They are:

* **Category 1 = sure**
* **Category 2 = somewhat sure**
* **Category 3 = not sure**

```{r virsorter-table}
virsorter_csv <- read.csv(virsorter_csv, comment.char = "#", header = FALSE, col.names = c("Contig_id", "Nb genes contigs", "Fragment", "Nb genes", "Category", "Nb phage hallmark genes", "Phage gene enrichment sig", "Non-Caudovirales phage gene enrichment sig", "Pfam depletion sig", "Uncharacterized enrichment sig", "Strand switch depletion sig", "Short genes enrichment sig"))
# Produce Table
virsorter_csv %>%
  arrange(Contig_id) %>%
  kable(caption = "Prophages annotated with phigaro software", escape = FALSE) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 12) %>%
  scroll_box(width = "1000px", height = "400px")
```

## PHAST