All prophage genes that were found in the query genome using PHAST database are showed in the Table below. In this table it is presented the query protein tag, its source prophage and gene function.

> Unfortunately, PHASTER database have no searchable interface to visualize its prophages. Therefore, this table has no links to external sources.

```{r phast-table}
# Create ID Column
gff$ID <- getAttributeField(gff$attributes, "ID", ";")

# To upper
gff$ID <- toupper(gff$ID)

# Remove duplicates based on bitscore
phast_blast <- phast_blast[order(phast_blast$qseqid, -abs(phast_blast$pident), -abs(phast_blast$bitscore) ), ]
phast_blast <-phast_blast[ !duplicated(phast_blast$qseqid), ]
phast_blast <- phast_blast[order(phast_blast$qseqid),]

# Produce Table
phast_blast$stitle <- gsub(phast_blast$stitle, pattern = " \\[", replacement = "\\|[")

# Data Complete
full_info <- merge.data.frame(phast_blast, gff, by.x = "qseqid", by.y = "ID")

# Create Position Column
full_info$position <- paste(full_info$seqname, ":", full_info$start, "-", full_info$end, sep = "")

full_info %>%
  select(qseqid, stitle, position) %>%
  separate(stitle, c("prophage", "id", "tag", "contig", "product", "species"), sep = "\\|") %>%
  select(prophage, species, qseqid, product, position) %>%
  arrange(prophage, position, species) %>%
  datatable(caption = "Prophage genes annotated using PHASTER database",
            colnames = c("Prophage Name", "Species source", "Query Protein ID", "Gene Product", "Query Protein Coordinates"),
          escape = FALSE,
          rownames = FALSE,
          options = list(pageLength = 5))
```
