```{r parsing-db-victors, eval=TRUE, include=FALSE, echo=FALSE}
# Download Victors Metadata
victors_meta <- read.delim("/work/victors/victors_metadata.tsv")

# Remove duplicates based on bitscore
victors_blast <- victors_blast[order(victors_blast$V1, -abs(victors_blast$V3), -abs(victors_blast$V13) ), ]
victors_blast <-victors_blast[ !duplicated(victors_blast$V1), ]
victors_blast <- victors_blast[order(victors_blast$V1),]

# Find Victors ID in each blast result
s <- strsplit(as.character(victors_blast$V2), "\\|")
y <- sapply(s, function(x) { x[2] })
victors_blast$DB_ID <- y

# Merge annotation with metadata
victors_full <- merge.data.frame(victors_blast, victors_meta, by.x = "DB_ID", by.y = "NCBI.Protein.GI")
victors_full <- victors_full[ !duplicated(victors_full$V1), ]

# Create ID Column
gff$ID <- getAttributeField(as.character(gff$attributes), "ID", ";")

# To upper
gff$ID <- toupper(gff$ID)

# Merge with GFF
victors_full <- merge.data.frame(victors_full, gff, by.x = "V1", by.y = "ID")

# Create Position Column
victors_full$position <- paste(victors_full$seqname, ":", victors_full$start, "-", victors_full$end, sep = "")

# Create DF for kable
blast <- victors_full %>% select(V1, Gene.Name, Protein.Name, Protegen.ID, position)

# Get urls
ids <- blast$Protegen.ID
urls <- lapply(ids, function(x){paste0("http://www.phidias.us/victors/gene_detail.php?c_mc_victor_id=", x)})

# Rename columns
colnames(blast) <- c("Query Protein ID", "Gene Name", "Product", "Database ID", "Query Protein Coordinates")
```

A summary of all the virulence factors which had at least on gene found in the query genome is shown in the Table below. The results are presented with gene name, gene product, genomics coordinates, query protein name and database ID. The annotation shown is linked to the database for further investigations.

```{r write-table-victors}
# Produce Table
## Subset dataframe
blast %>%
  mutate("Database ID" = cell_spec(`Database ID`, "html", link = urls)) %>%
  arrange(`Database ID`, `Query Protein ID`, `Gene Name`, `Product`) %>%
  datatable(caption = "Virulence Genes annotated from Victors database",
            colnames = c("Query Protein ID", "Gene Name", "Product", "Database ID", "Query Protein Coordinates"),
          escape = FALSE,
          rownames = FALSE,
          options = list(pageLength = 5))
```