```{r parsing-db-victors, eval=TRUE, include=FALSE, echo=FALSE}
# Download Victors Metadata
victors_meta <- read.delim("/work/victors/victors_metadata.tsv")

# Remove duplicates based on bitscore
victors_blast <- victors_blast[order(victors_blast$V1, -abs(victors_blast$V3), -abs(victors_blast$V13) ), ]
victors_blast <-victors_blast[ !duplicated(victors_blast$V1), ]
victors_blast <- victors_blast[order(victors_blast$V1),]

# Find Victors ID in each blast result
s <- strsplit(as.character(victors_blast$V2), "\\|")
y <- sapply(s, function(x) { x[2] })
victors_blast$DB_ID <- y

# Merge annotation with metadata
victors_full <- merge.data.frame(victors_blast, victors_meta, by.x = "DB_ID", by.y = "NCBI.Protein.GI")
victors_full <- victors_full[ !duplicated(victors_full$V1), ]

# Function to get Attribute Fields
getAttributeField <- function (x, field, attrsep = ";") { 
  s = strsplit(x, split = attrsep, fixed = TRUE) 
  sapply(s, function(atts) { 
    a = strsplit(atts, split = "=", fixed = TRUE) 
    m = match(field, sapply(a, "[", 1)) 
    if (!is.na(m)) { rv = a[[m]][2] 
    } 
    else { 
      rv = as.character(NA) 
    } 
    return(rv) 
  }) 
}

# Create ID Column
gff$ID <- getAttributeField(as.character(gff$V9), "id", ";")

# To upper
gff$ID <- toupper(gff$ID)

# Merge with GFF
victors_full <- merge.data.frame(victors_full, gff, by.x = "V1", by.y = "ID")

# Create Position Column
victors_full$position <- paste(victors_full$V1.y, ":", victors_full$V4.y, "-", victors_full$V5.y, sep = "")

# Create DF for kable
blast <- victors_full %>% select(V1, Gene.Name, Protein.Name, Protegen.ID, position)

# Get urls
ids <- blast$Protegen.ID
urls <- lapply(ids, function(x){paste0("http://www.phidias.us/victors/gene_detail.php?c_mc_victor_id=", x)})
```

All query genes that have a match in Victors database are described in Table \@ref(tab:write-table-victors). In order to make further investigations easier, all genes are linked to its matches in Victors database.

The results are presented with gene name, gene product, genomics coordinates, query protein name and database ID.

```{r write-table-victors}
# Produce Table
colnames(blast) <- c("Query Protein ID", "Gene Name", "Product", "Database ID", "Query Protein Coordinates")
blast %>%
  mutate("Database ID" = cell_spec(`Database ID`, "html", link = urls)) %>%
  arrange(`Query Protein ID`) %>%
  kable(caption = "Virulence Genes annotated from Victors database", escape = FALSE) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 12) %>%
  collapse_rows(columns = 1:5, valign = "middle")
```