```{r parsing-db, eval=TRUE, include=FALSE, echo=FALSE}
# Create ID Column
gff$ID <- getAttributeField(as.character(gff$attributes), "ID", ";")

# To upper
gff$ID <- toupper(gff$ID)

# Remove duplicates based on bitscore
vfdb_blast <- vfdb_blast[order(vfdb_blast$V1, -abs(vfdb_blast$V3), -abs(vfdb_blast$V13) ), ]
vfdb_blast <-vfdb_blast[ !duplicated(vfdb_blast$V1), ]
vfdb_blast <- vfdb_blast[order(vfdb_blast$V1),]

# Data Complete
full_info <- merge.data.frame(vfdb_blast, gff, by.x = "V1", by.y = "ID")

# Create Position Column
full_info$position <- paste(full_info$seqname, ":", full_info$start, "-", full_info$end, sep = "")

# Create DF for kable
blast <- full_info %>% select(V1, V14, position)
blast <- separate(blast, V14, c("VF ID", "other"), sep = " ", extra = "merge")
blast <- separate(blast, other, c("Gene Name", "other"), sep = " ", extra = "merge")
blast$other <- gsub(blast$other, pattern = " \\[", replacement = "#\\[")
blast <- separate(blast, other, c("Description", "Virulence Factor Name", "Subject Species"), sep = "#", extra = "merge")

# Sort
blast <- arrange(blast, `Virulence Factor Name`, `Gene Name`)

# Get urls
ids <- blast$`Virulence Factor Name`
ids <- str_match(ids, "VF.*[0-9]+")

urls <- lapply(ids, function(x){paste0("http://www.mgc.ac.cn/cgi-bin/VFs/vfs.cgi?VFID=", x, "#", x)})
```

A summary of all the virulence factors which had at least on gene found in the query genome is shown in the figure below (Figure \@ref(fig:vfs-png). Virulence genes are more detailed in Table below. It shows the gene product, source and name of each gene related to a Virulence Factor found in the query genome. The annotation shown is linked to the database for further investigations.

```{r vfs-png, fig.align='center', fig.show='hold', fig.cap="Virulence Factors annotated from [VFDB](http://www.mgc.ac.cn/VFs/main.htm) database", out.width="60%"}
summary <- blast %>%
  select("Virulence Factor Name", "Gene Name")
colnames(summary) <- c("Virulence_Factor_Name", "Gene_Name")
vfs <- plyr::count(summary, "Virulence_Factor_Name")
ggplot(vfs, aes(x=Virulence_Factor_Name, y=freq, fill=Virulence_Factor_Name)) + geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  xlab("Virulence Factors") +
  ylab("Number of annotated genes") +
  guides(fill=guide_legend(title = "Virulence Factors"))
```

### Supporting data

```{r write-table}
# Produce Table
## Rename Columns
colnames(blast) <- c("Query Protein ID", "VF ID", "Gene Name", "Description", "Virulence Factor Name", "Subject Species", "Query Protein Coordinates")
## Subset dataframe
blast %>%
  select("Virulence Factor Name", "VF ID", "Gene Name", "Description", "Subject Species", "Query Protein ID", "Query Protein Coordinates") %>%
  arrange(`Virulence Factor Name`, `VF ID`,`Gene Name`) %>%
  mutate("Virulence Factor Name" = cell_spec(`Virulence Factor Name`, "html", link = urls)) %>%
  datatable(caption = "Virulence Genes annotated from VFDB",
            colnames = c("Virulence Factor Name", "VF ID", "Gene Name", "Description", "Subject Species", "Query Protein ID", "Query Protein Coordinates"),
          escape = FALSE,
          rownames = FALSE,
          options = list(pageLength = 5))
```
