<style>
  .col2 {
    columns: 2 200px;         /* number of columns and width in pixels*/
    -webkit-columns: 2 200px; /* chrome, safari */
    -moz-columns: 2 200px;    /* firefox */
  }
  .col3 {
    columns: 3 100px;
    -webkit-columns: 3 100px;
    -moz-columns: 3 100px;
  }
</style>

Below it is summarized the result obtained with RGI. Roughly, it can be divided into two main categories: 

1. Perfect hits
    + detects perfect matches to the curated reference sequences and mutations in the CARD
2. Strict hits
    + detects previously unknown variants of known AMR genes, including secondary screen for key mutations, using detection models with CARD's curated similarity cut-offs to ensure the detected variant is likely a functional AMR gene

**Obs:** CARD RGI tool always tries to annotate functional AMR genes, however, depending on the assembly, a not functional gene may yet be annotated. There for, users are advised to double check genes annotated under Strict category.

### Perfect hits

<div class="col2">
```{r results='asis'}
## Name columns
colnames(rgi_tsv) <- c("Contig", "Start", "Stop", "Orientation", "Cut_Off", "Best_Hit_ARO", "ARO", "Drug Class", "Resistance Mechanism", "AMR Gene Family")
## Fix Contig Names
rgi_tsv$Contig <- gsub(pattern = "_[0-9]*$", replacement = "", rgi_tsv$Contig)
## Create Coordinate
rgi_tsv$Coordinates <- paste(rgi_tsv$Contig, ":", rgi_tsv$Start, "-", rgi_tsv$Stop, sep = '')
## Separate rows
rgi_tsv <- rgi_tsv %>%
  separate_rows(`Drug Class`, sep = "; ")

# Create List
string.list <- rgi_tsv %>% pull(6) %>% sort() %>% unique()
cat("- **AMR genes found in the query genome:**\n")
cat(paste("\t-", string.list), sep = '\n')

# Create List
string.list <- rgi_tsv %>% pull(8) %>% sort() %>% unique()
cat("- **Drug Classes targeted by the query genome:**\n")
cat(paste("\t-", string.list), sep = '\n')
```
</div>

#### Supporting Data

```{r rgi-resistance-genes}
# Create urls
urls <- lapply(rgi_tsv$ARO, function(x){paste0("https://card.mcmaster.ca/ontology/", x)})

# Produce Table
rgi_tsv %>%
  select("Resistance Mechanism", "Drug Class", "ARO", "Coordinates") %>%
  mutate("ARO" = cell_spec(ARO, "html", link = urls)) %>%
  group_by(`Resistance Mechanism`) %>%
  arrange(`Resistance Mechanism`, `Drug Class`) %>%
  kable(caption = "CARD resistance genes annotated and its target drug classes", escape = FALSE) %>%
  kable_styling(bootstrap_options = bt_opt, full_width = T, fixed_thead = T, font_size = 12) %>%
  collapse_rows(columns = 1:4, valign = "middle")
```

```{r rgi-resistome-png, fig.align='center', fig.show='hold', fig.cap="Resistome Predicted using CARD RGI tool", out.width="60%"}
rgi_count <- rgi_tsv %>%
  select(1,8)
colnames(rgi_count) <- c("Contig", "Class")
drugs <- plyr::count(rgi_count, "Class")
ggplot(drugs, aes(x=Class, y=freq, fill=Class)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  xlab("Targeted Drug Class") +
  ylab("Number of genes") +
  guides(fill=guide_legend(title = "Drug Classes"))
```

### Strict hits
