
#### Strict Hits

Strict hits genes are genes annotated with looser parameters and are not surely functional or correct. Which means, users should double check genes under this category.

```{r results='asis'}
# Create List
string.list <- rgi_strict %>% separate_rows(`Drug Class`, sep = "; ") %>% pull(`Drug Class`) %>% sort() %>% unique()
cat("- **Drug Classes targeted by the query genome:**\n")
cat(paste("\t-", string.list), sep = '\n')
```

```{r rgi-strict-resistance-genes}
# Produce Table
rgi_strict %>%
  select("Gene","Resistance Mechanism", "Drug Class", "ARO", "CARD Link", "Query Protein Coordinates", "Description") %>%
  group_by(`Resistance Mechanism`) %>%
  arrange(`Gene`, `CARD Link`,`Resistance Mechanism`, `Drug Class`) %>%
  datatable(caption = "CARD resistance genes annotated and its target drug classes",
            colnames = c("Gene","Resistance Mechanism", "Drug Class", "ARO", "CARD Link", "Genome Coordinates", "Description"),
          escape = FALSE,
          rownames = FALSE,
          options = list(pageLength = 5))
```

<br>

In the image below it is shown the amount of genes found to target these specific drug classes.

```{r rgi-strict-resistome-png, fig.align='center', fig.show='hold', fig.cap="Resistome Predicted using CARD RGI tool", out.width="80%"}
rgi_count <- rgi_strict %>%
  select(`Protein ID`,`Drug Class`)
colnames(rgi_count) <- c("Contig", "Class")
drugs <- plyr::count(rgi_count, "Class")
ggplot(drugs, aes(x=Class, y=freq, fill=Class)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  xlab("Targeted Drug Class") +
  ylab("Number of genes") +
  guides(fill=guide_legend(title = "Drug Classes"))
```