```{r parsing-db, eval=TRUE, include=FALSE, echo=FALSE}
ice_blast <- ice_blast[!grepl("complete chromossome", ice_blast$stitle),]
ice_blast <- ice_blast[order(ice_blast$qseqid, -abs(ice_blast$pident), -abs(ice_blast$bitscore) ), ]
ice_blast <-ice_blast[ !duplicated(ice_blast$qseqid), ]
ice_blast <- ice_blast[order(ice_blast$qseqid),]

# Create ID Column
gff$ID <- getAttributeField(gff$attributes, "ID", ";")

# To upper
gff$ID <- toupper(gff$ID)

# Data Complete
full_info <- merge.data.frame(ice_blast, gff, by.x = "qseqid", by.y = "ID")

# Create Position Column
full_info$position <- paste(full_info$seqname, ":", full_info$start, "-", full_info$end, sep = "")


blast <- full_info %>% select(qseqid, sseqid, stitle, position)
blast$name <- str_split(blast$stitle, pattern = "\\|", simplify = TRUE )[,6]
blast <- separate(blast, sseqid, c("Database", "ICE_id"), sep = "\\|")

# Get urls

urls <- lapply(blast$ICE_id, function(x){paste0("http://202.120.12.136:7913/ICEberg2/feature_page.php?ice_id=", x)})
```

All anotated genes from known ICEs are showed in the Table below. It is summarised the ICE id and all its genes that were found in the query genome. All of them are linked to the database for further investigations.

Users are advised to use the [ICEfinder](http://202.120.12.136:7913/ICEfinder/ICEfinder.html) tool to predict genomic position of known ICEs since we are not allowed to include this step under this pipeline.

```{r write-table}
# Produce Table
blast %>% select("ICE_id", "qseqid", "name", "position") %>%
  mutate("ICE_id" = cell_spec(ICE_id, "html", link = urls)) %>%
  arrange(ICE_id, position) %>%
  datatable(caption = "ICEs genes annotated from ICEberg database",
            colnames = c("ICE id", "Query Protein ID", "ICE's gene names", "Query Protein Coordinates"),
          escape = FALSE,
          rownames = FALSE,
          options = list(pageLength = 5))
```