Below it is summarized the results obtained with Resfinder Roughly, it can be divided into two main categories: 

1. Acquired resistance genes;
2. Point mutations conferring resistance.

Summary of the acquired resistance genes found in the genome:

```{r results='asis'}
# Read result
tab <- resfinder_tab

# Create List
string.list <- tab %>% pull(1) %>% unique()
cat("- **AMR genes found in the query genome:**\n")
cat(paste("\t-", string.list), sep = '\n')
```

## Acquired resistance genes

<caption>(#tab:resfinder-resistance-genes) Resistance genes annotated with Resfinder</caption>
```{r resfinder-resistance-genes}
resfinder_tab$Accession.no. <- resfinder_url(resfinder_tab$Accession.no.)
# Produce Table
resfinder_tab %>%
  datatable(escape = FALSE,
            filter = 'top',
            options = dt_opt_lst,
            extensions = 'Buttons',
            rownames = T)
```

```{r resfinder-resistome-png, fig.align='center', fig.show='hold', fig.cap="Resistome Predicted using NCBI's AMRFinderplus", out.width="45%"}
## Grab AMR
amr <- resfinder_tab %>%
  select(Contig, Phenotype)
classes <- lapply(amr$Phenotype, function(x) {
  strsplit(as.character(x), "resistance")[[1]][1] %>% str_trim()
  })
amr$Phenotype <- unlist(classes)
amr <- plyr::count(amr, "Phenotype")

## Plot
ggplot(amr, aes(x=Phenotype, y=freq, fill=Phenotype)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  xlab("") +
  ylab("Number of genes") +
  guides(fill=guide_legend(title = "Resistance Class"))
```

## Point mutations

<caption>(#tab:resfinder-pointfinder) Point mutations annotated with Resfinder</caption>
```{r resfinder-pointfinder}
# Produce Table
resfinder_point %>%
  datatable(escape = FALSE,
            filter = 'top',
            options = dt_opt_lst,
            extensions = 'Buttons',
            rownames = T)
```

```{r resfinder-pointfinder-png, fig.align='center', fig.show='hold', fig.cap="Resistome Predicted using Resfinder point mutations", out.width="45%"}
## Grab AMR
amr <- resfinder_point %>%
  select(Mutation, Resistance)
amr <- plyr::count(amr, "Resistance")

## Plot
ggplot(amr, aes(x=Resistance, y=freq, fill=Resistance)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  xlab("") +
  ylab("Number of point mutations mutations") +
  guides(fill=guide_legend(title = "Resistance Class"))
```
